File: app/src/main/java/com/semye/android/xml/ProductHandler.java
package com.semye.android.xml;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;


public class ProductHandler extends DefaultHandler {

    @Nullable
    private List<HashMap<String, String>> list = null;
    @Nullable
    private HashMap<String, String> hm = null;
    @Nullable
    private String s = null;

    @Override
    public void characters(char[] ch, int start, int length)
            throws SAXException {
        super.characters(ch, start, length);
        s = new String(ch, start, length);
    }

    @Override
    public void startDocument() throws SAXException {
        super.startDocument();
        list = new ArrayList<>();
    }

    @Override
    public void endDocument() throws SAXException {
        super.endDocument();
    }

    @Override
    public void startElement(String uri, @NonNull String localName, String qName, Attributes attributes) throws SAXException {
        super.startElement(uri, localName, qName, attributes);
        if (localName.equals("product")) {
            hm = new HashMap<>();
        }
    }

    @Override
    public void endElement(String uri, @NonNull String localName, String qName) throws SAXException {
        super.endElement(uri, localName, qName);
        if (hm != null && list != null) {
            if (localName.equals("id")) {
                hm.put("id", s);
            }
            if (localName.equals("name")) {
                hm.put("name", s);
            }
            if (localName.equals("price")) {
                hm.put("price", s);
            }
            if (localName.equals("product")) {
                list.add(hm);
            }
        }
    }

    @Nullable
    public List<HashMap<String, String>> getProduct() {
        return list;
    }


}

File: app/src/main/java/com/semye/android/xml/ProductHandler.kt
package com.semye.android.xml

import org.xml.sax.Attributes
import org.xml.sax.SAXException
import org.xml.sax.helpers.DefaultHandler
import java.util.*

class ProductHandler : DefaultHandler() {
    private var list: MutableList<HashMap<String, String?>>? = null
    private var hm: HashMap<String, String?>? = null
    private var s: String? = null
    @Throws(SAXException::class)
    override fun characters(ch: CharArray, start: Int, length: Int) {
        super.characters(ch, start, length)
        s = String(ch, start, length)
    }

    @Throws(SAXException::class)
    override fun startDocument() {
        super.startDocument()
        list = ArrayList()
    }

    @Throws(SAXException::class)
    override fun endDocument() {
        super.endDocument()
    }

    @Throws(SAXException::class)
    override fun startElement(
        uri: String,
        localName: String,
        qName: String,
        attributes: Attributes
    ) {
        super.startElement(uri, localName, qName, attributes)
        if (localName == "product") {
            hm = HashMap()
        }
    }

    @Throws(SAXException::class)
    override fun endElement(uri: String, localName: String, qName: String) {
        super.endElement(uri, localName, qName)
        if (hm != null && list != null) {
            if (localName == "id") {
                hm!!["id"] = s
            }
            if (localName == "name") {
                hm!!["name"] = s
            }
            if (localName == "price") {
                hm!!["price"] = s
            }
            if (localName == "product") {
                list!!.add(hm!!)
            }
        }
    }

    val product: List<HashMap<String, String?>>?
        get() = list
}